File: backend/src/main/java/com/tracex/backend/model/Defect.java
package com.tracex.backend.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "defects", indexes = {
    @Index(name = "idx_defects_board_id", columnList = "board_id")
})
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@ToString(exclude = "board")
@EqualsAndHashCode(exclude = "board")
public class Defect {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "board_id", nullable = false, foreignKey = @ForeignKey(name = "fk_defects_boards", value = ConstraintMode.CONSTRAINT))
    @JsonIgnore
    private Board board;

    @Column(nullable = false, length = 100)
    private String component; // e.g. "R34"

    @Column(nullable = false, length = 100)
    private String defect; // e.g. "Missing Component", "Solder Bridge"

    @Column(nullable = false, length = 50)
    private String severity; // "Critical", "High", "Medium", "Low"

    @Column(nullable = false)
    private double confidence; // 0.0 to 1.0

    @Column(name = "bounding_box", nullable = false, columnDefinition = "text")
    @Convert(converter = BoundingBoxConverter.class)
    private BoundingBox boundingBox;

    @Column(name = "image_path", length = 512)
    private String imagePath;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;
}

File: frontend/lib/models/defect.dart
class BoundingBox {
  final double x;
  final double y;
  final double width;
  final double height;

  BoundingBox({
    required this.x,
    required this.y,
    required this.width,
    required this.height,
  });

  factory BoundingBox.fromJson(Map<String, dynamic> json) {
    return BoundingBox(
      x: (json['x'] as num).toDouble(),
      y: (json['y'] as num).toDouble(),
      width: (json['width'] as num).toDouble(),
      height: (json['height'] as num).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'x': x,
      'y': y,
      'width': width,
      'height': height,
    };
  }
}

class Defect {
  final String id;
  final String boardId;
  final String component;
  final String defect;
  final String severity; // "Critical", "High", "Medium", "Low"
  final double confidence;
  final BoundingBox boundingBox;
  final String? imagePath;
  final DateTime createdAt;

  Defect({
    required this.id,
    required this.boardId,
    required this.component,
    required this.defect,
    required this.severity,
    required this.confidence,
    required this.boundingBox,
    this.imagePath,
    required this.createdAt,
  });

  factory Defect.fromJson(Map<String, dynamic> json) {
    return Defect(
      id: json['id'] as String,
      boardId: json['board_id'] as String,
      component: json['component'] as String,
      defect: json['defect'] as String,
      severity: json['severity'] as String,
      confidence: (json['confidence'] as num).toDouble(),
      boundingBox: BoundingBox.fromJson(json['bounding_box'] as Map<String, dynamic>),
      imagePath: json['image_path'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'board_id': boardId,
      'component': component,
      'defect': defect,
      'severity': severity,
      'confidence': confidence,
      'bounding_box': boundingBox.toJson(),
      'image_path': imagePath,
      'created_at': createdAt.toIso8601String(),
    };
  }
}
